src/network/networkctl.c

```diff
@@ -1395,87 +1395,71 @@ static int dump_dhcp_leases(Table *table, const char *prefix, sd_bus *bus, const
-        if (r < 0) {
-                bool quiet = sd_bus_error_has_name(&error, SD_BUS_ERROR_UNKNOWN_PROPERTY);
-
-                log_full_errno(quiet ? LOG_DEBUG : LOG_WARNING,
-                               r, "Failed to query link DHCP leases: %s", bus_error_message(&error, r));
-                return 0;
-        }
-
-        r = sd_bus_message_enter_container(reply, 'v', "a(uayayayayt)");
+        r = bus_call_method(bus, bus_network_mgr, "DescribeLink", &error, &reply, "i", link->ifindex);
         if (r < 0)
-                return bus_log_parse_error(r);
+                return log_error_errno(r, "Failed to get link description: %s", bus_error_message(&error, r));
 
-        r = sd_bus_message_enter_container(reply, 'a', "(uayayayayt)");
+        r = sd_bus_message_read(reply, "s", &text);
         if (r < 0)
                 return bus_log_parse_error(r);
 
-        while ((r = sd_bus_message_enter_container(reply, 'r', "uayayayayt")) > 0) {
-                _cleanup_free_ char *id = NULL, *ip = NULL;
-                const void *client_id, *addr, *gtw, *hwaddr;
-                size_t client_id_sz, sz;
-                uint64_t expiration;
-                uint32_t family;
-
-                r = sd_bus_message_read(reply, "u", &family);
-                if (r < 0)
-                        return bus_log_parse_error(r);
-
-                r = sd_bus_message_read_array(reply, 'y', &client_id, &client_id_sz);
-                if (r < 0)
-                        return bus_log_parse_error(r);
-
-                r = sd_bus_message_read_array(reply, 'y', &addr, &sz);
-                if (r < 0 || sz != 4)
-                        return bus_log_parse_error(r);
-
-                r = sd_bus_message_read_array(reply, 'y', &gtw, &sz);
-                if (r < 0 || sz != 4)
-                        return bus_log_parse_error(r);
-
-                r = sd_bus_message_read_array(reply, 'y', &hwaddr, &sz);
-                if (r < 0)
-                        return bus_log_parse_error(r);
+        r = json_parse(text, 0, &interface, NULL, NULL);
+        if (r < 0)
+                return log_error_errno(r, "Failed to parse JSON: %m");
 
-                r = sd_bus_message_read_basic(reply, 't', &expiration);
-                if (r < 0)
-                        return bus_log_parse_error(r);
+        dhcp_server = json_variant_by_key(interface, "DHCPServer");
+        if (dhcp_server) {
+                JSON_VARIANT_ARRAY_FOREACH(lease, json_variant_by_key(dhcp_server, "Leases")) {
+                        JsonVariant *hostname = NULL;
+                        _cleanup_free_ uint8_t *client_id = NULL, *addr = NULL;
+                        _cleanup_free_ char *id = NULL, *ip = NULL;
+                        size_t client_id_sz, addr_sz;
+                        uint64_t expiration_usec;
+
+                        r = json_variant_bytes(
+                                        json_variant_by_key(lease, "ClientId"), &client_id, &client_id_sz);
```

Comment:
instead of doing repeated json_variant_by_key() calls I think it would be much better to use json_dispatch() which write the various fields to a struct. It's a lot more efficient, has better error handling, and also prettier to use. You find a lot of examples in our codebase which already do that.

Reply:
Implemented this. I put the struct definition into the function (I saw it done in other places in the code, so I think it's fine). LMK if you don't like it :)